meta/BRIP-0003.md

```diff
@@ -0,0 +1,150 @@
+  if req.Height < sbtEnableHeight {
+    return timeoutCommit
+  }
+
+  // Case 2: current height == SBTEnableHeight => initialize the block delay.
+  if req.Height == sbtEnableHeight {
+    s.blockDelay = &BlockDelay{
+      InitialTime:       req.Time,
+      InitialHeight:     req.Height,
+      PreviousBlockTime: req.Time,
+    }
+    return timeoutCommit
+  }
+
+  // Case 3: current height > SBTEnableHeight => compute next block delay
+  if s.blockDelay != nil {
+    return s.blockDelay.Next(req.Time, req.Height)
+  }
+
+  panic(fmt.Sprintf(
+      "SBT not initialized at height %d (past enable height %d). State sync is not supported",
+      req.Height,
+      sbtEnableHeight,
+  ))
+}
+
+func (d *BlockDelay) Next(curBlockTime time.Time, curBlockHeight int64) time.Duration {
+  // We can't use 0 delay as that would make CometBFT read the timeout_commit from config file
+  const noDelay = 1 * time.Microsecond
+
+  // Reset the initial time and height if the time between blocks is greater
+  // than MaxDelayBetweenBlocks. This makes the current time and height the
+  // initial one as if the upgrade happened just now.
+  if curBlockTime.Sub(d.PreviousBlockTime) > MaxDelayBetweenBlocks {
+    d.InitialTime = curBlockTime
+    d.InitialHeight = curBlockHeight
+    d.PreviousBlockTime = curBlockTime
+    return TargetBlockTime
+  }
+
+  d.PreviousBlockTime = curBlockTime
+
+  expectedTime := d.InitialTime.Add(TargetBlockTime * time.Duration(curBlockHeight - d.InitialHeight))
+  if curBlockTime.Before(expectedTime) {
+    return expectedTime.Sub(curBlockTime)
+  }
+  return noDelay
+}
+```
+
+To prevent rapid block generation after a prolonged halt (e.g., due to downtime or upgrades), the internal checkpoint is reset after `MaxDelayBetweenBlocks` - the maximum delay between two consecutive blocks. `MaxDelayBetweenBlocks` is set to `5 minutes`, which balances distinguishing real halts from network congestion while avoiding excessive catch-up.
+
+To make this work, BeaconKit keeps track of `PreviousBlockTime`. BeaconKit must persist this, along with `initialTime` and `initialHeight`, for every block (on Commit) to ensure crash recovery.
+
+### State Sync Support
+
+Currently, SBT does not support [state sync](https://docs.cometbft.com/v1.0/explanation/core/state-sync). If a node syncs from a height greater than SBTEnableHeight, BeaconKit lacks access to the latest initialTime. In this version, such scenarios must panic:
+
+Future versions may introduce support for persisting and recovering SBT state from snapshots.
+
+## Rationale
+
+SBT leverages [Predictable Block Times](https://github.com/cometbft/cometbft/blob/main/docs/references/architecture/adr-115-predictable-block-times.md) feature from CometBFT.
+
+## Backward Compatibility
+
+A hard fork is required to enable these changes. This change does not affect execution clients, so it is fully backward-compatible with existing wallets and applications.
```

Comment:
Can we add a small description why SBT cannot be implemented at a timestamp (like our hardforks do) but must be activated at a block height specifically?